README.md
# Patterns

A starterkit to create web styleguides with [Fractal](http://fractal.build/) and [Webpack](https://webpack.js.org/).

- Fractal pre-configured with [Nunjucks](https://mozilla.github.io/nunjucks/)
- JavaScript bundling with Webpack and [Babel](http://babeljs.io/)
- Sass compilation, including [Autoprefixer](https://github.com/postcss/autoprefixer)
- Manage design tokens with [Style Dictonary](https://amzn.github.io/style-dictionary/#/README)
- [SVG icons sprite](https://css-tricks.com/svg-symbol-good-choice-icons/) generation
- Live reload (with hot module replacement) for comfortable development
- Automated release management with [release-it](https://github.com/webpro/release-it)


## Installation

Prerequisites: [Node.js](https://nodejs.org/en/) >= 10, <= 12

Inside the directory of your choice, install a copy of the starterkit with:

```bash
curl -L https://github.com/ingvoo/patterns/archive/master.tar.gz | tar zx --strip 1
```

Then install the npm dependencies with:

```bash
npm install
```


## Getting started

To start the development server, run:

```bash
npm start
```

You can now access your styleguide at [localhost:3000](http://localhost:3000).

You’re all set, start to:

- Create components as `.nunj` (Nunjucks) files inside the `components` directory
- Write some style inside `assets/scss/main.scss`
- Write JavaScript inside `assets/scripts/main.js`
- Put some `*.svg` icons in the `assets/icons` directory
- Write documentation as `.md` (Markdown) files inside the `docs` directory.


## Build

You can build a static version of the styleguide to deploy it wherever you like by running:

```
npm run build
```

The generated files go to the `dist` directory.


## Release

The starterkit comes with a preconfigured release management tool. It will automatically update the `CHANGELOG.md` file at the root of the project based on the commit messages as long as they follow the [Angular commit guidelines](https://github.com/angular/angular.js/blob/master/DEVELOPERS.md#-git-commit-guidelines). It will also bump the version number in the `package.json`, run the build command above, commit, tag and push the changes. This process is interactive and you’ll be able to skip steps manually if you like.

To release a new version run:

```bash
npm run release [patch|minor|major|version_number]
```

By default the version bump is automatically determined based on the commits messages.

Read more in the [release-it documentation](https://github.com/webpro/release-it).


## Misc

### Browsers support

The browsers support is defined in `.browserslistrc`. It’s used both by [Autoprefixer](https://github.com/postcss/autoprefixer) for the CSS and by [@babel/preset-env](https://babeljs.io/docs/en/babel-preset-env) for the JavaScript.

Check [browserslist’s documentation](https://github.com/ai/browserslist) to change the browser support.

### Icons

Icons placed inside `assets/icons` are combined into a sprite called `icons.svg` when referenced. To load them, either reference them in your JavaScript:

```js
import 'icons/foo.svg';
```

Or in your stylesheet:

```scss
background-image: url('../icons/foo.svg');
```

Webpack will automatically rewrite the links to the sprite with the correct identifier. See [svg-sprite-loader](https://github.com/kisenka/svg-sprite-loader) documentation for more information.

You can then easily use an icon in a template with the icon snippet:

```nunj
{% render '@icon', { id: 'foo', class: 'bar' } %}
```

**Warning:** This method of including remote SVG file is not supported by Internet Explorer 11 and below. You may want to polyfill it with [svgxuse](https://github.com/Keyamoon/svgxuse).

## Credits
This project was originally cloned from https://github.com/liip/styleguide-starterkit. Very special thanks to to [Liip](https://github.com/liip) and to [Benoît Burgener](https://github.com/LeBenLeBen).


CHANGELOG.md
# [0.10.0](https://github.com/ingvoo/patterns/compare/0.9.0...0.10.0) (2022-12-06)


### Features

* make breakpoints available via json tokens ([4b619cd](https://github.com/ingvoo/patterns/commit/4b619cd9561e3090a4817ecb6a5382ba20242ad5)), closes [#68](https://github.com/ingvoo/patterns/issues/68)
* set sass vars in favor of custom properties ([775e06a](https://github.com/ingvoo/patterns/commit/775e06ab615b0f1da1988da7d4b6d7d3e0f34780))
* setup missing flex directions properties ([13b46ac](https://github.com/ingvoo/patterns/commit/13b46ac51f54d004c38fff8aae9365689da7453e))

# [0.9.0](https://github.com/ingvijonasson/patterns/compare/0.8.0...0.9.0) (2022-10-18)


### Features

* add typography and more missing basic design tokens ([b0b0081](https://github.com/ingvijonasson/patterns/commit/b0b0081dd11b0bfaf3609b366756c6e446e171f3))
* introduce typography mixings ([614f13d](https://github.com/ingvijonasson/patterns/commit/614f13dc9b45c46e7bed2214708e0f5113a16858))
* use scss variables in favor of custom properties for theming ([d4d02a4](https://github.com/ingvijonasson/patterns/commit/d4d02a4f8547228ba0f2018d7fe8caffc319ca42))

# [0.8.0](https://github.com/ingvijonasson/patterns/compare/0.7.0...0.8.0) (2022-04-03)


### Features

* add breakpoints tokens ([aa5aab6](https://github.com/ingvijonasson/patterns/commit/aa5aab6e9bba9abf7651fce9263445a4b542306c))
* introduce color index documentation ([76cbcb0](https://github.com/ingvijonasson/patterns/commit/76cbcb0e8b5638d40d4fe7c030321c9332fb244f))
* update sass version ([181d544](https://github.com/ingvijonasson/patterns/commit/181d544531ece16f8dba370ec165749701c3d299))

# [0.7.0](https://github.com/ingvijonasson/patterns/compare/0.6.0...0.7.0) (2022-01-11)


### Bug Fixes

* update text-black color value ([a578d6e](https://github.com/ingvijonasson/patterns/commit/a578d6e0e70dabd527c5f880e7906031d2865041))


### Features

* add documentation for spaces ([7ea1dc3](https://github.com/ingvijonasson/patterns/commit/7ea1dc35cc89b61f2bbe208bd41c21e29ac6f64c))
* add json token output to public directory for later generating styleguide ([0f70a9e](https://github.com/ingvijonasson/patterns/commit/0f70a9e218f03affc5b118f30e034f4f301c168d))
* add script and tokens for style-directive ([38904c7](https://github.com/ingvijonasson/patterns/commit/38904c71a09432a649622da23d76d82febaf799f))
* introduce color documentations page for tokens ([dff3d01](https://github.com/ingvijonasson/patterns/commit/dff3d01a6c1d632fc3ecf578b183012c0c9357b2))
* remove all scss color variables in favor of custom properties ([b98d626](https://github.com/ingvijonasson/patterns/commit/b98d626e7f5af9db4657fdcbf41c993f7123bd5b))
* use spaces custom properties in favor of scss size function ([aed6ff7](https://github.com/ingvijonasson/patterns/commit/aed6ff750e5381f65f295d50911972e703da3297))

# [0.6.0](https://github.com/ingvijonasson/patterns/compare/0.5.0...0.6.0) (2021-05-25)


### Features

* introduce accordion component ([91e897e](https://github.com/ingvijonasson/patterns/commit/91e897e9f0801c029c345e698299e9c4f948996c))
* introduce component collapse ([48080e6](https://github.com/ingvijonasson/patterns/commit/48080e61c907732e1a5b7df1d18c6cb0469d677b))

# [0.5.0](https://github.com/ingvijonasson/patterns/compare/0.4.2...0.5.0) (2021-04-20)


### Features

* introduce form-radio and form-label ([fdb202d](https://github.com/ingvijonasson/patterns/commit/fdb202dee2aea1dc5f35d6cb807dd04fec24af46))

## [0.4.2](https://github.com/ingvijonasson/patterns/compare/0.4.1...0.4.2) (2021-04-04)

## [0.4.1](https://github.com/ingvijonasson/patterns/compare/0.4.0...0.4.1) (2020-10-17)

# [0.4.0](https://github.com/ingvijonasson/patterns/compare/0.2.0...0.4.0) (2020-10-16)


### Features

* add btn secondary, small and large modifications ([c9254e4](https://github.com/ingvijonasson/patterns/commit/c9254e4f12dce080af1932be6ee65116a6ce77dc))
* add display utilities ([9c4547f](https://github.com/ingvijonasson/patterns/commit/9c4547fb30d22f95226e5a6f70b3880d05ac6d1b))
* add responsive margin utilities ([c9f9f58](https://github.com/ingvijonasson/patterns/commit/c9f9f5804cdc61cf8c40af4095d4f90b692c0b1a))
* introduce border utilities ([56364e7](https://github.com/ingvijonasson/patterns/commit/56364e74b5e09d5f1934c24ec2f48e863d911a0c))
* introduce embed object for embeding iframes ([af0699c](https://github.com/ingvijonasson/patterns/commit/af0699c31820e12733ef456058dacd390bb34b0a))
* introduce flex utilities ([a50bf11](https://github.com/ingvijonasson/patterns/commit/a50bf1198d9fc24abe4e136e858ee7e5c093c979))
* introduce form-input ([6dcb793](https://github.com/ingvijonasson/patterns/commit/6dcb7934ff313f5732eeb74b00fd29a6c78923db))
* introduce grid utilities ([7643499](https://github.com/ingvijonasson/patterns/commit/76434994fadec4b742c1ffb2d18ce28152a1894b))
* introduce height and width utilities ([4af8032](https://github.com/ingvijonasson/patterns/commit/4af80329672f76ac003e6489e85173af2f52fe63))
* introduce list and list-description ([24c26bc](https://github.com/ingvijonasson/patterns/commit/24c26bc9df2ee77ed8c377b85f94e2280a83c993))
* introduce padding utilities ([515b897](https://github.com/ingvijonasson/patterns/commit/515b8975732a09b855665dab94c0956292c937e4))
* introduce position utility classes ([ef6ff9f](https://github.com/ingvijonasson/patterns/commit/ef6ff9f355c72d7ee6c1b29b3811e3a6f939a1e4))
* introduce text utilities ([e14bc5f](https://github.com/ingvijonasson/patterns/commit/e14bc5f5d95513b587f83d0582c5680060cefd49))

# [0.3.0](https://github.com/ingvijonasson/patterns/compare/0.2.0...0.3.0) (2020-10-16)


### Features

* add btn secondary, small and large modifications ([c9254e4](https://github.com/ingvijonasson/patterns/commit/c9254e4f12dce080af1932be6ee65116a6ce77dc))
* add display utilities ([9c4547f](https://github.com/ingvijonasson/patterns/commit/9c4547fb30d22f95226e5a6f70b3880d05ac6d1b))

fractal.js
const path = require('path');

const fractal = module.exports = require('@frctl/fractal').create();
const pkg = require(path.join(__dirname, 'package.json'));


/*-------------------------------------------------------*\
  Feel free to adapt Fractal config below to your needs
\*-------------------------------------------------------*/

/**
 * Metadata
 */
fractal.set('project.title', 'Patterns');
fractal.set('project.version', pkg.version);

/**
 * Files location
 */
fractal.components.set('path', path.join(__dirname, 'components'));
fractal.docs.set('path', path.join(__dirname, 'docs'));
fractal.web.set('static.path', path.join(__dirname, 'public'));

/**
 * Build options
 */
// If you change the build destination, you should adapt webpack.common.js "output.path" too.
fractal.web.set('builder.dest', 'dist');

/**
 * Templating
 */
// Use Nunjucks as the template engine
fractal.components.engine('@frctl/nunjucks');
fractal.docs.engine('@frctl/nunjucks');
// Look for templates with a ".nunj" extension
fractal.components.set('ext', '.nunj');


/*----------------------------------------*\
  Change the following at your own risk
\*----------------------------------------*/

/**
 * Server configuration
 */
fractal.web.set('server.port', 4000);
fractal.web.set('server.sync', true);

/**
 * Prevent Bluebird warnings like "a promise was created in a handler but was not returned from it"
 * caused by Nunjucks from polluting the console
 */
const bluebird = require('bluebird');
bluebird.config({
  warnings: false
});


fractal.config.js
const path = require('path');

const fractal = (module.exports = require('@frctl/fractal').create());
const pkg = require(path.join(__dirname, 'package.json'));

/*-------------------------------------------------------*\
  Feel free to adapt Fractal config below to your needs
\*-------------------------------------------------------*/

/**
 * Metadata
 */
fractal.set('project.title', 'Patterns');
fractal.set('project.version', pkg.version);

/**
 * Files location
 */
fractal.components.set('path', path.join(__dirname, 'components'));
fractal.docs.set('path', path.join(__dirname, 'docs'));
fractal.web.set('static.path', path.join(__dirname, 'public'));

/**
 * Build options
 */
// If you change the build destination, you should adapt webpack.common.js "output.path" too.
fractal.web.set('builder.dest', 'dist');

/**
 * Templating
 */
// Use Nunjucks as the template engine
fractal.components.engine('@frctl/nunjucks');
fractal.docs.engine('@frctl/nunjucks');
// Look for templates with a ".nunj" extension
fractal.components.set('ext', '.nunj');

/*----------------------------------------*\
  Change the following at your own risk
\*----------------------------------------*/

/**
 * Server configuration
 */
fractal.web.set('server.port', 4000);
fractal.web.set('server.sync', true);

/**
 * Prevent Bluebird warnings like "a promise was created in a handler but was not returned from it"
 * caused by Nunjucks from polluting the console
 */
const bluebird = require('bluebird');
bluebird.config({
  warnings: false,
});


webpack.config.js
const path = require('path');
const webpack = require('webpack');
const MiniCssExtractPlugin = require('mini-css-extract-plugin');
const SpriteLoaderPlugin = require('svg-sprite-loader/plugin');

module.exports = {
  mode: process.env.NODE_ENV,
  resolve: {
    modules: [
      path.resolve(__dirname, 'assets/scripts'),
      path.resolve(__dirname, 'assets'),
      'node_modules'
    ],
    extensions: ['.js']
  },
  entry: {
    main: path.resolve(__dirname, 'assets/scripts/main.js'),
  },
  output: {
    path: path.resolve(__dirname, 'dist'),
    publicPath: '/',
    filename: '[name].js'
  },
  module: {
    rules: [
      {
        test: /\.js$/,
        exclude: /node_modules/,
        loader: 'babel-loader',
      },
      {
        test: /\.scss$/,
        use: [
          {
            loader: MiniCssExtractPlugin.loader,
            options: {
              hmr: process.env.NODE_ENV === 'development',
            }
          },
          'css-loader',
          {
            loader: 'postcss-loader',
            options: {
              plugins: [
                require('autoprefixer')(),
                require('cssnano')(),
              ],
            },
          },
          {
            loader: 'sass-loader',
            options: {
              sassOptions: {
                includePaths: ['node_modules'],
              }
            },
          },
        ],
      },
      {
        test: /\.(svg|png|jpe?g|gif|woff|woff2|eot|ttf|otf)$/,
        exclude: path.resolve('./assets/icons'),
        use: [
          {
            loader: 'file-loader',
            options: {
              name: '[name].[ext]',
              outputPath: 'assets/',
            }
          }
        ]
      },
      {
        test: /\.svg$/,
        include: path.resolve('./assets/icons'),
        use: [
          {
            loader: 'svg-sprite-loader',
            options: {
              extract: true,
              spriteFilename: 'assets/icons.svg',
              esModule: false,
            }
          },
          'svgo-loader'
        ],
      }
    ]
  },
  plugins: [
    new MiniCssExtractPlugin(),
    new SpriteLoaderPlugin(),
  ],
  devServer: {
    historyApiFallback: true,
    compress: true,
    proxy: {
      '**': 'http://localhost:4000'
    },
    port: 3000,
    stats: {
      colors: true
    },
    overlay: true,
  },
  optimization: {
    splitChunks: {
      cacheGroups: {
        styles: {
          name: 'styles',
          test: /\.css$/,
          chunks: 'all',
          enforce: true,
        },
      },
    },
  },
};
